src/publishers/publishers/ftp_publisher.py

```diff
@@ -49,32 +43,32 @@ def get_prefix_and_ftp(ftp_server_url):
             f.write(bytes_data)
             f.close()
 
-            ftp_prefix = get_prefix_and_ftp(ftp_url)[0]
-            ftp_url = get_prefix_and_ftp(ftp_url)[1]
-
             ftp_data = urlsplit(ftp_url)
 
-            ftp_hostname = ftp_data.scheme + '.' + ftp_data.hostname
+            ftp_hostname = ftp_data.hostname
             ftp_username = ftp_data.username
             ftp_password = ftp_data.password
-            ftp_port = ftp_data.port
 
             remote_path = ftp_data.path + filename
 
-            if ftp_prefix == 'sftp:':
+            if ftp_data.scheme == 'sftp':
+                ssh_port = ftp_data.port if ftp_data.port else 22
                 ssh = paramiko.SSHClient()
                 ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-                ssh.connect(ftp_hostname, ftp_port, ftp_username, ftp_password)
+                ssh.connect(hostname=ftp_hostname, port=ssh_port, username=ftp_username, password=ftp_password)
                 sftp = ssh.open_sftp()
                 sftp.put(filename, remote_path)
                 sftp.close()
-                os.remove(filename)
-            else:
+            elif ftp_data.scheme == 'ftp':
+                ftp_port = ftp_data.port if ftp_data.port else 21
                 ftp = ftplib.FTP()
-                ftp.connect(ftp_hostname, ftp_port)
+                ftp.connect(host=ftp_hostname, port=ftp_port)
                 ftp.login(ftp_username, ftp_password)
                 ftp.storbinary('STOR ' + remote_path, open(filename, 'rb'))
                 ftp.quit()
-                os.remove(filename)
+            else:
+              raise Exception("Schema '{}' not supported, choose 'ftp' or 'sftp'".format(ftp_data.scheme))
         except Exception as error:
             BasePublisher.print_exception(self, error)
+        finally:
+            os.remove(filename)
```

Comment:
Initially I thought `finally` is not a good place for os.remove(), because it wipes the file even if the upload was not successful. That's why in the original version it was inline, and also the reason why someone thought temporary local file names should be predictable (to allow troubleshooting).

However on the other read of our code, we should probably handle this better upstream, and also prevent a race condition with naming at the same time. I'll open an issue for that, and accept this code so the failed uploads don't clutter the container forever.

Reply:
Or use `else:` in the try-except clause

Reply:
we could have just left it as it was, but I like this approach, followed by more cleanup which I propose in https://github.com/SK-CERT/Taranis-NG/issues/66